### DISCUSSION TITLE
Optical Beta

### DISCUSSION THREAD

**Discussion by erichelgeson** [erichelgeson/BlueSCSI-v1#150] [Show and tell]
This is a beta release of optical support for BlueSCSI based on #130 by @mynameistroy

You can now create SCSI optical drives with `.iso`, `.bin`, `.toast`, and other raw optical formats. 

BlueSCSI will try to determine the correct block size for optical, and default to 2048 if none can be determined.

## Reporting Success, Issues, and Questions in beta 

If you have Success, Issues, and Questions please use the linked discussion below.

If you have technical feedback please report that in the Pull Request.

## Usage:

Follows the same naming conventions as Hard Disks, except using `CD`

`CD30_2048-MacOS8.6.iso` - SCSI ID 3, LUN 0, 2048 sector size.

Remember file names must be less than 32 characters.

### Bootable Images

Some images created do not have the proper bits to be detected as bootable on a Mac. We are compiling of known good images here https://github.com/erichelgeson/BlueSCSI/wiki/CD-Image-Support

## Known behaviors

Eject currently does nothing, we have some interesting plans around CD switching in the works.

If an image file has the read only attribute checked it is currently skipped (some CD images have this checked) - will be fixed in next beta. 

### USB Boot loader 

If you have the USB Boot loader you can flash any of the `-USB` firmware to your devices using this branch of the [BlueSCSI-Updater](https://github.com/erichelgeson/Bluescsi-updater/tree/optical-beta) or [QMK Toolbox ](https://github.com/qmk/qmk_toolbox)(generic DFU util for windows/mac)

### ST-Link

If you do not have a USB bootloader you can flash via [ST-Link v2.](https://github.com/erichelgeson/BlueSCSI/wiki/Flashing-(Advanced))

<hr /><em>This discussion was created from the release <a href='https://github.com/erichelgeson/BlueSCSI/releases/tag/v1.1-JOKKER-OPTICAL-20220705'>Optical Beta</a>.</em>

**Comment by xnop** [comment]
More CD images of MacOS 7.6.1, 8.1, Legacy on

http://quadra605.mcdonnelltech.com

**Reply by erichelgeson** [reply]
Have you verified any of these boot? Still looking for a 7.6.1 that is bootable.

**Reply by erichelgeson** [reply]
That 7.6.1 you added to the wiki page will not boot on my Q950 - it actually just locks up the machine.

**Reply by xnop** [reply]
Hi, I was able to reinstall my LC475 aka Q605 from that image, and same for the MacOS 8.1.

**Reply by xnop** [reply]
I could double check with my developer DVDs which contain all versions of MacOS 7.x, 8.x and 9.x in several languages…

**Comment by ghost** [comment]
I'm attempting to install A/UX using a BlueSCSI to mount the install CD toast image. The A/UX installer is very picky on hardware. What does the CD drive hardware look like to the OS?

**Reply by erichelgeson** [reply]
It identifies as a vendor: "BLUESCSI", product: "CD-ROM CDU-55S", version: "1.9a" - which is enough to fool atleast the normal apple cdrom drivers.

**Reply by ghost** [reply]
Any possibility of making this configurable in a future release?

**Reply by erichelgeson** [reply]
For sure - we currently have scsi-info.txt but now with multiple device types it'll be a bit more tricky, but we'll come up with something. If there was a common case you had in mind I could build a custom bin for you.

**Reply by ghost** [reply]
I was able to get around my immediate need. When you get to a point where you need someone to test an A/UX install from CD let me know and I can test. There are a few compatible drives. 

I'm not sure if this is a common occurrence with the A/UX installation program or not, or just an issue with the particular drive I was using (CDU561-25 Sony Apple 300), the CD had to be inserted only after the installer prompted for a disk. If I booted with the CD in the drive, it did not recognize it when needed. I imagine this could be difficult to implement while the BlueSCSI is in use. There may need to be an on-the-fly method of mounting and ejecting disks.

**Comment by D0uglas082** [comment]
_PBREG->CRH = 0x11111111*DB_MODE_IN shows warning (integer overflow).  Should " * " be changed ?_


(BlueSCSI.h)
```c++
// Put DB and DP in output mode
#define SCSI_DB_OUTPUT() { PBREG->CRL=(PBREG->CRL &0xfffffff0)|DB_MODE_OUT; PBREG->CRH = 0x11111111 * DB_MODE_OUT; }
// Put DB and DP in input mode
#define SCSI_DB_INPUT()  { PBREG->CRL=(PBREG->CRL &0xfffffff0)|DB_MODE_IN ; PBREG->CRH = (uint32_t)(0x11111111 * DB_MODE_IN); }
```

**Reply by mynameistroy** [reply]
The cast to uint32_t should be properly handling that, and I don't get that error when I build it. Are you building the optical branch or the mainline branch?

**Reply by D0uglas082** [reply]
I built the optical branch and got build warning. I wonder about setting pbreg->crh (GPIO config ?) with overflow value is ok or not.

**Reply by mynameistroy** [reply]
Yeah I only ask because I don't see that error when I build it on my machine. So I'm not entirely sure why you are.

**Comment by lausvi** [comment]
I tried to attach the Myst.toast_.sit from https://macintoshgarden.org/games/myst as SCSI ID6. On my MacOS8 machine, the System Profiler can see a volume called Myst, but it's not getting mounted.

**Reply by erichelgeson** [reply]
I used Download `#1` (the iso) and it worked great. Not all toast images will be compatible unfortunately, but I'll give this one a look.

**Reply by lausvi** [reply]
Ok, I tried the first listed ISO as well, but still no dice. Using Transoft SCSI Director, I can see the CD device (Bluescsi cd-rom cdu-55s); the device information shows it as removable, size: 524 MB, Partitions N/A, Map status: Major problems. Attempting to mount it in SCSI director gives no erros but it doesn't appear on desktop.

BlueSCSI's log shows it deceted the image as follows: 
- CD60-Myst.iso CDROM / 551219200bytes / 538300KiB / 525MiB MODE2:0 BlockSize:2048

HOWEVER; a Mac OS 8.1 ISO worked straight away, so perhaps there is some issue with the Myst images?

**Comment by zefrenchtoon** [comment]
Did you unstuff the file first to restore it as an iso ? 🤔

**Reply by lausvi** [reply]
I extracted the .toast (out of the .sit), yes. I didn't try the other versions.

**Comment by zefrenchtoon** [comment]
Is this image a multi-session one ? (Joliet/iso+hfs)
If yes, maybe it is a clue to check ?
